src/presentation.ts

```diff
@@ -310,6 +325,118 @@ async function createPresentation<R extends PresentationRequest>(
   };
 }
 
+async function preparePresentation<R extends PresentationRequest>({
+  request,
+  context: walletContext,
+  credentials,
+}: {
+  request: R;
+  context: WalletContext<R>;
+  credentials: (StoredCredential & { key?: string })[];
+}): Promise<{
+  context: Field;
+  fieldsToSignStringArray: string[];
+  credentialsUsed: Record<string, StoredCredential>;
+  clientNonce: Field;
+  compiledRequest: {
+    program: Program<Output<R>, Inputs<R>>;
+    verificationKey: VerificationKey;
+  };
+}> {
+  let { program, verificationKey } = await Presentation.compile(request);
+  let clientNonce = Field.random();
+  let context = request.deriveContext(request.inputContext, walletContext, {
+    clientNonce,
+    vkHash: verificationKey.hash,
+    claims: hashClaims(request.claims),
+  });
+
+  let credentialsNeeded = Object.entries(request.spec.inputs).filter(
+    (c): c is [string, CredentialSpec] => c[1].type === 'credential'
+  );
+  let credentialsUsed = pickCredentials(
+    credentialsNeeded.map(([key]) => key),
+    credentials
+  );
+  let credentialsAndTypes = credentialsNeeded.map(([key, credentialType]) => {
+    let credentialAndType = { ...credentialsUsed[key]!, credentialType };
+    if (isSubclass(credentialType.data, DynamicRecord.Base)) {
+      let { owner, data } = credentialAndType.credential;
+      credentialAndType.credential = {
+        owner,
+        data: credentialType.data.from(data),
+      };
+    }
+    return credentialAndType;
+  });
+
+  let credHashes = credentialsAndTypes.map(({ credentialType, credential }) =>
+    hashCredential(credentialType.data, credential)
+  );
+  let issuers = credentialsAndTypes.map(({ credentialType, witness }) =>
+    credentialType.issuer(witness)
+  );
+
+  const fieldsToSign = [
+    context,
+    ...zip(
+      credHashes.map((c) => c.hash),
+      issuers
+    ).flat(),
+  ];
+  const msg = fieldsToSign.toString();
+  return {
+    context,
+    fieldsToSignStringArray: fieldsToSign.map((f) => f.toString()),
```

Comment:
nit: could you rename this to `message` or `messageFields`? The information that it's a string array is important, but already part of the type